### DISCUSSION TITLE
Challenge 8 Solution *** SPOILERS INCLUDED ***

### DISCUSSION THREAD

**Discussion by npa528** [Cyfrin/foundry-full-course-cu#126] [General]
I open a new topic about Challenge 8 so that we can discuss our different ways to solve this challenge. I make it clear in the title that below we will discuss solution for this challenge, so any users who want to give it a try, **SHOULD NOT** continue reading this topic.

I would like also to see this for all Challenges. Include "SPOILERS INCLUDED" (or similar) to prepare any users from reading the solution, so the rest of us can discuss different approaches so that we can learn also from each other (decentralized after all) and not only from @PatrickAlphaC (hehe). 

I violated this rule in previous topic about challenge 8 where I provided almost all details of my approach to solve challenge. (https://github.com/Cyfrin/foundry-full-course-f23/discussions/102) so I apologise for this.

So I took my friend's @alymurtazamemon advise from that previous discussion, and try to check another way of solving this. And he was right (partially).

I should have paid more attention to the data of the transaction, which I ingored.

```
Error: cannot estimate gas; transaction may fail or may require manual gas limit (error={"code":-32000,"message":"execution reverted"}, method="estimateGas", transaction={"from":"0x05E3Bd725678352E24d79D2266e12129EdB69474","to":"0x76B50696B8EFFCA6Ee6Da7F6471110F334536321","data":"**0x29a30a7000000000000000000000000000000000000000000000000000000000075bcd15000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000066667646764660000000000000000000000000000000000000000000000000000**","accessList":null}, code=UNPREDICTABLE_GAS_LIMIT, version=providers/5.1.2)
```

Checking on the chapters of this tutorial, I found that there is a chapter called "Verifying Metamask Transactions" in Lesson 11: Foundry NFTs | MoodNFT

There, at point Decode the calldata ([timestamp 10:33:50](https://youtu.be/sas02qSFZ74?t=38030)) is explained the command 
``` cast --calldata-decode``` which did the trick.

I am still confused as of why Challenge 8 (from Lesson 8) needs knowledge that **will be tought** in Lesson 11 to be solved but OK, It is my fault that I did not pay attention to data field of transaction and did not do my own research of how to handle this data.

And most important, I still haven't figured out - and here is the part where I would like to hear your solutions guys - 
how we jumped from contract 0x76B50696B8EFFCA6Ee6Da7F6471110F334536321 where was the target contract the IPFS webpage was communicating with, to contract 0xf215a0b6DD88d6029b5385D6fab51968337E963D where is the method "solveChallenge()" for Challenge8.

**Comment by cromewar** [comment]
You are amazing @npa528 this is the way to create discussions, I would love to see opinions from the community in this regard.

**Comment by RichardPalladino** [comment]
I don't believe this requires `cast`, though I still haven't completely solved it, I can see how to solve it.

[The IPFS link](https://ipfs.io/ipfs/QmUgc1xpNgTMwSF2GBS5FTFKmSdrmSnLJD5DoYZtcQ2n5X/) shows the index webpage

reading the source code, that imports [index.js](https://ipfs.io/ipfs/QmUgc1xpNgTMwSF2GBS5FTFKmSdrmSnLJD5DoYZtcQ2n5X/index.js)

that imports [constants.js](https://ipfs.io/ipfs/QmUgc1xpNgTMwSF2GBS5FTFKmSdrmSnLJD5DoYZtcQ2n5X/constants.js)

In constants.js we find:

1. Contract addresses for both Sepolia (0x76B50696B8EFFCA6Ee6Da7F6471110F334536321) and Arbitrum (0x39338138414Df90EC67dC2EE046ab78BcD4F56D9); 
2. The additional `secretValue` that needs to be sent with the `solveChallenge` contract call; and
3. The contract's ABI

Okay.  So there are several ways to do this.  
- You could call the contract function using cast.  (I'm not familiar enough to do this so I'm not sure if this is the route I'll take.)
- You could make an interface and call the contract using something like Remix.  (My initial attempt to make an interface and call the `solveChallenge()` function still has failed, so I'm not sure if I'll continue to pursue this route.)
- You could be a dick and ruin it for everyone else by using the ABI to verify the contract on Etherscan, then make the function call from there. (DON'T DO THIS!)

You're talking about where 0x76B50696B8EFFCA6Ee6Da7F6471110F334536321 comes from when clearly the Sepolia contract is 0xf215a0b6DD88d6029b5385D6fab51968337E963D?  I recommend you check out constants.js .

Clearly the challenge is to find that the interface is janked up and we need to create our own way to interface the contracts -- correctly.

**Comment by RichardPalladino** [comment]
... I just read your #102 discussion comment
 
> I violated this rule in previous topic about challenge 8 where I provided almost all details of my approach to solve challenge. (#102) so I apologise for this.

...
 
> I am still confused as of why Challenge 8 (from Lesson 8) needs knowledge that **will be tought** in Lesson 11 to be solved but OK, It is my fault that I did not pay attention to data field of transaction and did not do my own research of how to handle this data.

... It doesn't _need_ cast.  There are several ways to interface the contract correctly.  You could have even re-worked the interface with web developer tools.  Some students did use cast successfully, but it's not needed.
 
> And most important, I still haven't figured out - and here is the part where I would like to hear your solutions guys - how we jumped from contract 0x76B50696B8EFFCA6Ee6Da7F6471110F334536321 where was the target contract the IPFS webpage was communicating with, to contract 0xf215a0b6DD88d6029b5385D6fab51968337E963D where is the method "solveChallenge()" for Challenge8.

So here is the deal: the QR code gives a link, but the actual contracts we need are listed in the [NFT factory](https://sepolia.etherscan.io/token/0x76b50696b8effca6ee6da7f6471110f334536321#readContract).  If you got to that link (for Sepolia, since that's where you seem to be operating from) and enter 7 as the input for `getChallengeContract`, you'll find the ACTUAL contract that should have been used. (The 0 index is for challenge 1, so 7 index is for challenge 8 contract.)

My above post stands.  The challenge is to see that the interface code is messed up, and we have to find a way to interface the _correct contract_, correctly.

First you gotta know that the contract in the interface code is incorrect.  You do that by getting the correct contract address from the NFT factory as I describe above. (See my comment in discussion topic #135 .)

Once you do that, the ABI code obtained from the IPFS link in the QR code will let you know how to interface the contract, and it also has the secret value needed to pass the challenge.  

You can interface the contract any way you want.  The `cast` tool certainly can be used but, as you noted, we haven't learned that much.  I ended up just writing a quick interface with Remix and interfacing the correct contract that way.  (I passed the Arbitrum challenge.  I'm going to do it again on Sepolia just to practice cast, because I want to learn that tool better too.)

**Reply by 78carla** [reply]
Hi @RichardPalladino,
firstly I have completed the challenge using the `cast` cheatcode. For me, it is much easier than using the interfaces. But thank your message I jumped also using the interfaces. I struggled a bit (with the IERC721Receiver interface) but in the end, I got it. Happy to have read this. Available to help if anyone has issues. Thanks for this thread,

**Comment by Teoaks** [comment]
Am I the only one not beeing able to see anything when opening the link from the QR code, or is that part of the challenge?

**Reply by RichardPalladino** [reply]
I think it may be you.  Others have had similar issues.  (Search the discussions, including closed / answered items.)

For your awareness, you may check out #135

**Comment by yipu3** [comment]
Here is my process of solving this challenge.

1. Right click -> Inspect -> Source. There is source code of the web page.
2. I found the address of [the called contract](https://arbiscan.io/address/0x39338138414Df90EC67dC2EE046ab78BcD4F56D9#code) in `constants.js`. After looking over the code of this contract I realized that it is not a challenge contract because it doesn't has a `solveChange` function. It is actually the NFT contract of all challenges!
3. So, how can we find the correct contract address of challenge 8? There is exactly a `getChallengeContract` interface in the NFT contract! Because it is 0-indexed to read the 8th challenge I called this interface with a parameter 7
<img width="452" alt="Screen Shot 2024-02-23 at 16 22 39" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/5365604/453afced-f02b-4ab3-855b-b36c121cf3a8">

4. Now that I had found the challenge address. I replaced the contract address called in `index.js`.
<img width="1031" alt="Screen Shot 2024-02-23 at 16 27 39" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/5365604/5af71a02-443d-4ef8-b7b9-4279fd69bb47">

6. Then I managed to call the `solveChallenge` via the Solve button.

**Reply by cqlyj** [reply]
OMG I have never thought in this way though I know some js...... YOU are definitely genius!

**Reply by EngrPips** [reply]
Go crush it hahaha.

**Reply by xueyuanhuang** [reply]
@yipu3 you are genius, I am trying to modify `constants.js` but it didn't works? I am guessing that's because it will not reimport `constants.js` again in `index.js` when I calling transaction.

currently it has a problem as the github repo changing the hint as:
```
IPFS hash for website - ipfs://QmdCBj5B188qzXgG1LUCD6aKe6RJHqbSgu7YfPyncDirzh
```
but it seems like the IPFS hash doesn't exist....
